### PR TITLE
Ready for review.

### PR DESCRIPTION
**Pull request by vdkrasny**
- npm run start
- npm run searchTearm
- npm run leaderboard

### CODE REVIEW

File: src/DataBaseClient.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+                    .then(() => console.log(item.joke))
+                    .catch(err => console.log(`DBClient.write Error: ${err}`));
+            })
+            .catch(err => console.log(`DBClient.read Error: ${err}`));
+    }
+
+    findMostСommon(key) {
+        this.read()
+            .then((parsedData) => {
+                if (parsedData[key] !== undefined && parsedData[key].length !== 0) {
```
Comment (BSiarhei) [inline src/DataBaseClient.js]: you can simplify

File: db/DataBaseClient.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+                .then(() => {
+                    fs.readFile(this.filePath, (err, data) => {
+                        if (err) {
+                            reject(err);
+                        }
+
+                        try {
+                            resolve(JSON.parse(data));
+                        } catch (e) {
+                            resolve([]);
```
Comment (BSiarhei) [inline db/DataBaseClient.js]: Why do you return [] if data of your file is not valid?

File: db/DataBaseClient.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        });
+    }
+
+    checkFileExist() {
+        return new Promise((resolve, reject) => {
+            fs.stat(this.filePath, (err) => {
+                if (err === null) {
+                    resolve();
+                } else if (err.code === 'ENOENT') {
+                    this.write([])
```
Comment (BSiarhei) [inline db/DataBaseClient.js]: Why do you do this? Your method is named checkFileExist. There is nothing about creating.

File: src/httpsRequestTo.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
+const https = require('https');
```
Comment (BSiarhei) [inline src/httpsRequestTo.js]: httpsRequestTo.js To is redundant

File: services/jokeService.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+    const firstPageRawData = await httpsRequestTo(apiUrl);
+    const {
+        current_page: currentPage,
+        next_page: nextPage,
+        total_pages: totalPages,
+        results: firstPageJokes
+    } = JSON.parse(firstPageRawData);
+
+    if (currentPage < nextPage && nextPage <= totalPages) {
```
Comment (BSiarhei) [inline services/jokeService.js]: why nextPage?

File: src/httpsRequestTo.js (hunk 4 above)
Comment (vdkrasny) [reply src/httpsRequestTo.js]: Does it have to be a part of the service?

File: db/DataBaseClient.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+            return this.write(parsedData);
+        } catch (err) {
+            return console.log(`Error: ${err.message}`);
+        }
+    }
+
+    read() {
+        return new Promise(async (resolve, reject) => {
+            try {
```
Comment (BSiarhei) [inline db/DataBaseClient.js]: unnecessary

File: db/DataBaseClient.js [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+                reject(err);
+            }
+        });
+    }
+
+    write(data) {
+        return new Promise((resolve, reject) => {
+            fs.writeFile(this.filePath, JSON.stringify(data), (err) => {
+                if (err) {
+                    reject(err);
```
Comment (BSiarhei) [inline db/DataBaseClient.js]: a best practice is to add return

File: services/jokeService.js [hunk 8]
```diff
@@ -0,0 +1,118 @@
+const https = require('https');
+
+const DataBaseClient = require('../db/DataBaseClient');
+
+const { getRandomInt } = require('../utils');
+
+const fileName = 'jokes';
+const jokesDBClient = new DataBaseClient(fileName);
+
+function apiRequest(url) {
```
Comment (BSiarhei) [inline services/jokeService.js]: should be moved out

File: services/jokeService.js [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+        headers: { 'Accept': 'application/json' },
+        method: 'GET'
+    };
+
+    return new Promise((resolve, reject) => {
+        https
+            .request(url, options, (response) => {
+                const body = [];
+
+                response.on('data', chunk => body.push(chunk));
```
Comment (BSiarhei) [inline services/jokeService.js]: there is no error handler for response

File: services/jokeService.js [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+        .sort((a, b) => idsWithQuantity[b] - idsWithQuantity[a])[0];
+
+    return jokes.find(({ id }) => id === mostCommonId);
+}
+
+module.exports.getJokeByTerm = async function (term) {
+    try {
+        const jokes = await getJokes({ term });
+        const randomIdx = getRandomInt(0, jokes.length);
+        const { id, joke } = jokes[randomIdx];
```
Comment (BSiarhei) [inline services/jokeService.js]: it is strange to have property joke in joke

File: services/jokeService.js [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+        await jokesDBClient.add({ id, joke });
+
+        return joke;
+    } catch (err) {
+        return err;
```
Comment (BSiarhei) [inline services/jokeService.js]: return error?

File: services/jokeService.js [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+};
+
+module.exports.getMostCommonJoke = async function () {
+    try {
+        const jokes = await jokesDBClient.read();
+        const { joke } = findMostCommonJoke(jokes);
+
+        return joke;
+    } catch (err) {
+        return err;
```
Comment (BSiarhei) [inline services/jokeService.js]: return error?

File: services/jokeService.js (hunk 10 above)
Comment (vdkrasny) [reply services/jokeService.js]: @BSiarhei but there is **jokes**

File: db/DataBaseClient.js [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+            .then((parsedData) => {
+                if (!(parsedData instanceof Array)) {
+                    throw new Error(`No data was saved. Please, check the ${this.filePath} file structure. It must be an array.`);
+                }
+
+                parsedData.push(data);
+
+                return this.write(parsedData);
+            })
+            .catch(({ message }) => console.log(`Error: ${message}`));
```
[the remark this answers is not in the document]
Comment (vdkrasny) [reply db/DataBaseClient.js]: @BSiarhei could you explain what should I do in this case?

**Pull request merged**